frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,1124 @@
+		) -> DispatchResult {
+			T::SuperMajorityOrigin::ensure_origin(origin)?;
+			let candidate = T::Lookup::lookup(candidate)?;
+			ensure!(Self::is_candidate(&candidate), Error::<T, I>::NotCandidate);
+			ensure!(!Self::is_member(&candidate), Error::<T, I>::AlreadyMember);
+
+			Self::remove_candidate(&candidate)?;
+			if let Some(deposit) = DepositOf::<T, I>::take(&candidate) {
+				T::Slashed::on_unbalanced(T::Currency::slash_reserved(&candidate, deposit).0);
+			}
+
+			Self::deposit_event(Event::CandidateRejected { candidate });
+			Ok(())
+		}
+
+		/// Elevate an ally to fellow.
+		#[pallet::weight(T::WeightInfo::reject_candidate())]
+		pub fn elevate_ally(
+			origin: OriginFor<T>,
+			ally: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			T::SuperMajorityOrigin::ensure_origin(origin)?;
+			let ally = T::Lookup::lookup(ally)?;
+			ensure!(Self::is_ally(&ally), Error::<T, I>::NotAlly);
+			ensure!(!Self::is_votable_member(&ally), Error::<T, I>::AlreadyElevated);
+
+			Self::remove_member(&ally, MemberRole::Ally)?;
+			Self::add_member(&ally, MemberRole::Fellow)?;
+
+			Self::deposit_event(Event::AllyElevated { ally });
+			Ok(())
+		}
+
+		/// As a member, retire and back to an ordinary account and unlock its deposit.
+		#[pallet::weight(T::WeightInfo::retire())]
+		pub fn retire(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(!Self::is_kicking(&who), Error::<T, I>::KickingMember);
+
+			let role = Self::member_role_of(&who).ok_or(Error::<T, I>::NotMember)?;
+			Self::remove_member(&who, role)?;
+			let deposit = DepositOf::<T, I>::take(&who);
+			if let Some(deposit) = deposit {
+				let err_amount = T::Currency::unreserve(&who, deposit);
+				debug_assert!(err_amount.is_zero());
+			}
+			Self::deposit_event(Event::MemberRetired { member: who, unreserved: deposit });
+			Ok(())
+		}
+
+		/// Kick a member to ordinary account with its deposit slashed.
+		#[pallet::weight(T::WeightInfo::kick_member())]
+		pub fn kick_member(
+			origin: OriginFor<T>,
+			who: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			T::SuperMajorityOrigin::ensure_origin(origin)?;
+			let member = T::Lookup::lookup(who)?;
+			ensure!(Self::is_kicking(&member), Error::<T, I>::NotKickingMember);
+
+			let role = Self::member_role_of(&member).ok_or(Error::<T, I>::NotMember)?;
+			Self::remove_member(&member, role)?;
+			let deposit = DepositOf::<T, I>::take(member.clone());
+			if let Some(deposit) = deposit {
+				T::Slashed::on_unbalanced(T::Currency::slash_reserved(&member, deposit).0);
+			}
+			Self::deposit_event(Event::MemberKicked { member, slashed: deposit });
+			Ok(())
+		}
+
+		/// Add accounts or websites into blacklist.
+		#[pallet::weight(T::WeightInfo::add_blacklist(infos.len() as u32, T::MaxWebsiteUrlLength::get()))]
+		pub fn add_blacklist(
+			origin: OriginFor<T>,
+			infos: Vec<BlacklistItemOf<T, I>>,
+		) -> DispatchResult {
```

Comment:
There are a few spots in this pallet where naming/docs are confusing. This is one example, the function name `add_blacklist` makes it sound like it's adding a list, when it's really adding items to lists. Also the arg `infos` is strange as it's not a word. Would suggest something like `fn add_items_to_blacklist(origin, items)`.

Rather than point out everything in comments here, I'd be happy to make a PR on this branch with suggested changes.

Reply:
Looked more in depth and came across quite a bit of docs that are wrong and more APIs/events/errors that are confusing. I've started a branch, will make a PR to this.

Reply:
https://github.com/xlc/substrate/pull/2